radixdlt-core/radixdlt/src/main/java/com/radixdlt/network/p2p/proxy/ProxyCertificateManager.java

```diff
@@ -0,0 +1,309 @@
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Manages the proxy certificates:
+ *  - issued by this node
+ *  - granted to this node
+ *  - received from other peers
+ */
+public final class ProxyCertificateManager {
+	private static final Logger log = LogManager.getLogger();
+
+	private final NodeId self;
+	private final P2PConfig config;
+	private final int networkId;
+	private final Addressing addressing;
+	private final ECKeyOps ecKeyOps;
+	private final PeersView peersView;
+	private final PeerControl peerControl;
+	private final AddressBook addressBook;
+	private final RemoteEventDispatcher<GrantedProxyCertificate> grantedProxyCertificateDispatcher;
+	private final RemoteEventDispatcher<ProxyCertificatesAnnouncement> proxyCertificatesAnnouncementDispatcher;
+	private final Object lock = new Object();
+
+	// the certificates that were granted to this node
+	private final Map<NodeId, ProxyCertificate> receivedProxyCertificates = new ConcurrentHashMap<>();
+
+	// the certificates issued by this node
+	private final Map<NodeId, ProxyCertificate> issuedProxyCertificates = new ConcurrentHashMap<>();
+
+	// the certificates received from other nodes (granted to them)
+	private final Map<NodeId, Set<VerifiedProxyCertificate>> verifiedProxies = new ConcurrentHashMap<>();
+
+	@Inject
+	public ProxyCertificateManager(
+		@Self NodeId self,
+		P2PConfig config,
+		@NetworkId int networkId,
+		Addressing addressing,
+		ECKeyOps ecKeyOps,
+		PeersView peersView,
+		PeerControl peerControl,
+		AddressBook addressBook,
+		RemoteEventDispatcher<GrantedProxyCertificate> grantedProxyCertificateDispatcher,
+		RemoteEventDispatcher<ProxyCertificatesAnnouncement> proxyCertificatesAnnouncementDispatcher
+	) {
+		this.self = Objects.requireNonNull(self);
+		this.config = Objects.requireNonNull(config);
+		this.networkId = networkId;
+		this.addressing = Objects.requireNonNull(addressing);
+		this.ecKeyOps = Objects.requireNonNull(ecKeyOps);
+		this.peersView = Objects.requireNonNull(peersView);
+		this.peerControl = Objects.requireNonNull(peerControl);
+		this.addressBook = Objects.requireNonNull(addressBook);
+		this.grantedProxyCertificateDispatcher = Objects.requireNonNull(grantedProxyCertificateDispatcher);
+		this.proxyCertificatesAnnouncementDispatcher = Objects.requireNonNull(proxyCertificatesAnnouncementDispatcher);
+	}
+
+	public ImmutableSet<ProxyCertificate> getReceivedProxyCertificates() {
+		synchronized (lock) {
+			return ImmutableSet.copyOf(receivedProxyCertificates.values());
+		}
+	}
+
+	public ImmutableSet<NodeId> getVerifiedProxiesForNode(NodeId nodeId) {
+		synchronized (lock) {
+			return verifiedProxies.getOrDefault(nodeId, Set.of())
+				.stream()
+				.filter(cert -> cert.expiresAt() > Time.currentTimestamp())
+				.map(VerifiedProxyCertificate::grantee)
+				.collect(ImmutableSet.toImmutableSet());
+		}
+	}
+
+	public void handlePeerConnected(PeerEvent.PeerConnected peerConnected) {
+		synchronized (lock) {
```

Comment:
`lock` necessary?

Reply:
It's not needed right now, but I'd keep it anyway. Might result in hard to spot bugs if someone later modifies one of the methods used here.